### ISSUE TITLE
sqlsmith bug hunt: 'internal error: entered unreachable code', src/frontend/src/optimizer/plan_node/batch_nested_loop_join.rs:50:23

### ISSUE THREAD

**Issue by marvenlee2486** [singularity-data/risingwave#4220]
```
running 1 test
Panic on SQL:
WITH with_0 AS (SELECT SMALLINT '11' AS col_0, sum(DISTINCT (40 << ((INT '23' >> (SMALLINT '23' | SMALLINT '19')) % INT '37')) * t_12.p_retailprice) AS col_1 FROM supplier AS t_1, m6 AS t_2, m7 AS t_3, hop(bid, bid.date_time, INTERVAL '1', INTERVAL '35872') AS hop_4, m2 AS t_5, m9 AS t_6, m1 AS t_7, m5 AS t_8 JOIN m9 AS t_9 ON t_8.col_2 = t_9.col_2, m1 AS t_10, part AS t_11, part AS t_12, m9 AS t_13 WHERE false GROUP BY hop_4.price, t_12.p_brand, hop_4.bidder, t_1.s_phone, t_6.col_2, t_12.p_size, t_1.s_address, t_1.s_comment, t_13.col_2, t_11.p_container, t_6.col_0, hop_4.date_time, t_3.col_0, t_11.p_retailprice HAVING true) SELECT t_14.col_1 # 6 AS col_0 FROM with_0, m0 AS t_14, tumble(bid, bid.date_time, INTERVAL '395740') AS tumble_15, partsupp AS t_16, m6 AS t_17, part AS t_18, orders AS t_19, orders AS t_20, nation AS t_21, lineitem AS t_22, part AS t_23, partsupp AS t_24 GROUP BY t_21.n_regionkey, t_18.p_comment, t_22.l_linestatus, t_18.p_name, t_21.n_nationkey, t_22.l_shipinstruct, t_24.ps_comment, t_18.p_retailprice, t_19.o_orderkey, t_24.ps_partkey, t_20.o_shippriority, t_20.o_custkey, t_22.l_extendedprice, t_14.col_0, t_22.l_partkey, t_22.l_shipdate, t_24.ps_suppkey, t_22.l_quantity, tumble_15.auction, t_19.o_shippriority, t_20.o_comment, tumble_15.bidder, t_20.o_clerk, t_23.p_comment, t_20.o_orderstatus, t_18.p_brand, t_16.ps_comment, t_23.p_retailprice, t_16.ps_suppkey, t_14.col_2, t_22.l_shipmode, t_22.l_commitdate, t_16.ps_availqty, tumble_15.date_time, t_23.p_brand, t_18.p_mfgr, t_24.ps_availqty, t_18.p_size, t_19.o_orderdate, t_20.o_orderdate, t_19.o_comment, t_19.o_orderstatus, t_19.o_orderpriority, t_18.p_type, t_14.col_1, t_23.p_container, t_22.l_discount, t_18.p_partkey, t_19.o_totalprice, t_16.ps_partkey, t_16.ps_supplycost, t_20.o_orderpriority, t_18.p_container, t_20.o_totalprice, t_24.ps_supplycost HAVING false ORDER BY t_24.ps_supplycost ASC
Reason:
panicked at 'internal error: entered unreachable code', src/frontend/src/optimizer/plan_node/batch_nested_loop_join.rs:50:23
test run_sqlsmith_on_frontend_18 ... FAILED
```

**Comment by marvenlee2486** [comment]
https://github.com/singularity-data/risingwave/issues/3989

this is solved, but then 
when gen_agg and gen_complex_query (gen with) is on together, this error occurs , 

Not only this, https://github.com/singularity-data/risingwave/issues/4218 occurs when this happens also

**Comment by Graphcalibur** [comment]
I'm unable to reproduce this error. The test `run_sqlsmith_on_frontend_18` passes for me on the latest version of the main branch. Perhaps this was also fixed by #4230?

**Comment by marvenlee2486** [comment]
You need to enable agg first because currently it is workaround. see `src/tests/sqlsmith/src/expr.rs` line 78 and 79, 
uncomment 78 and comment 79.

then use 
`./risedev test run_sqlsmith_on_frontend`
It may not be _18 one cuz currently sqlsmith does not always produce the same result.

**Comment by marvenlee2486** [comment]
use this command instead
`./risedev test run_sqlsmith_on_frontend --features enable_sqlsmith_unit_test`

**Comment by marvenlee2486** [comment]
```
Panic on SQL:
WITH with_0 AS (SELECT 10 AS col_0, REAL '55.915434335112614' AS col_1 FROM m5 AS t_3, m3 AS t_4, orders AS t_5, part AS t_6, m9 AS t_7, m5 AS t_10, m2 AS t_11, person AS t_12, m8 AS t_13 JOIN lineitem AS t_14 ON t_13.col_0 = t_14.l_shipmode, orders AS t_15, m0 AS t_16, partsupp AS t_17 WHERE true GROUP BY t_15.o_orderkey, t_12.credit_card, t_15.o_orderdate, t_12.state, t_17.ps_availqty, t_12.city, t_6.p_partkey, t_12.id, t_12.date_time, t_12.email_address, t_5.o_totalprice, t_17.ps_comment, t_17.ps_partkey, t_4.col_2, t_5.o_orderdate, t_5.o_clerk, t_15.o_orderpriority, t_5.o_custkey, t_3.col_0, t_10.col_0, t_12.name, t_15.o_orderstatus, t_17.ps_suppkey, t_15.o_shippriority, t_17.ps_supplycost, t_6.p_retailprice, t_15.o_totalprice, t_6.p_name, t_6.p_type, t_6.p_size HAVING min(DISTINCT true)) SELECT 14 | INT '62' AS col_0, 45.57166333787481 AS col_1, TIMESTAMP '2022-07-29 02:48:54' AS col_2 FROM with_0, tumble(bid, bid.date_time, INTERVAL '223376') AS tumble_18, m6 AS t_19, m7 AS t_20, m8 AS t_21, m0 AS t_22, person AS t_23, m8 AS t_24, part AS t_25, m4 AS t_26, m7 AS t_27, m8 AS t_30, m4 AS t_31, m3 AS t_32 GROUP BY t_21.col_2, t_19.col_1, t_25.p_brand, t_23.credit_card, t_20.col_1, t_30.col_0, t_23.state, t_24.col_2, t_27.col_0, tumble_18.bidder, t_24.col_3, tumble_18.date_time, t_22.col_1, t_19.col_0, t_25.p_container, t_19.col_3, t_24.col_0, t_25.p_mfgr, t_30.col_1, t_30.col_3, t_25.p_retailprice, t_25.p_size, t_22.col_0, t_20.col_0, t_23.date_time HAVING t_22.col_0 LIMIT 39
Reason:
panicked at 'internal error: entered unreachable code', src/frontend/src/optimizer/plan_node/batch_nested_loop_join.rs:51:23
```

```
running 1 test
Panic on SQL:
WITH with_0 AS (SELECT DATE '2022-07-20' AS col_0, INTERVAL '150507' * FLOAT '65.14696747832866' AS col_1, max(DISTINCT t_10.s_nationkey) + t_6.l_receiptdate AS col_2 FROM m7 AS t_1, region AS t_2, region AS t_3 JOIN m6 AS t_4 ON t_3.r_regionkey = t_4.col_1, m9 AS t_5, lineitem AS t_6, customer AS t_9, supplier AS t_10, m0 AS t_11 JOIN m0 AS t_12 ON t_11.col_0 = t_12.col_0, m6 AS t_13 JOIN m7 AS t_14 ON t_13.col_3 = t_14.col_3, partsupp AS t_15, part AS t_16 GROUP BY t_1.col_1, t_1.col_2, t_6.l_commitdate, t_9.c_mktsegment, t_5.col_0, t_16.p_partkey, t_10.s_address, t_9.c_custkey, t_16.p_retailprice, t_6.l_tax, t_6.l_quantity, t_15.ps_comment, t_1.col_3, t_9.c_name, t_6.l_returnflag, t_6.l_discount, t_16.p_mfgr, t_15.ps_partkey, t_6.l_receiptdate, t_6.l_extendedprice, t_2.r_comment, t_6.l_partkey, t_6.l_suppkey) SELECT FLOAT '80.01304248696479' AS col_0 FROM with_0, nation AS t_17, auction AS t_18, auction AS t_19 JOIN part AS t_20 ON t_19.item_name = t_20.p_container, tumble(person, person.date_time, INTERVAL '294939') AS tumble_21, lineitem AS t_22, person AS t_25, m0 AS t_28, tumble(person, person.date_time, INTERVAL '648857') AS tumble_29, bid AS t_30 WHERE INT '8' < REAL '26.436809264699555' GROUP BY t_30.price, t_18.seller ORDER BY t_18.seller ASC, t_18.seller DESC, t_30.price ASC
Reason:
panicked at 'internal error: entered unreachable code', src/frontend/src/optimizer/plan_node/batch_nested_loop_join.rs:51:23
```

**Comment by marvenlee2486** [comment]
I found a simplified version of it,
It happens when two condition is meet
1) distinct agg is used in subquery (doesn,t matter it is FROM subquery or WITH clause)
2) When there is other relation outside the query

Possible reason: Idk, this bug pretty strange to me.

```
dev=> create table t1 ( x int , y int);
CREATE_TABLE
dev=> create table t2 (x int, y int);
CREATE_TABLE

```
When agg distinct is used in a simple query - no problem
```
dev=> select min(distinct 1) from t1;
 ?column? 
----------
         
(1 row)

dev=> select min(distinct x) from t1;
 ?column? 
----------
         
(1 row)
```

When distinct agg is used in subquery, no problem
```
with w as (select min (distinct 1) from t1) select 1 from w;
 ?column? 
----------
        1
(1 row)

dev=> with w as (select min (distinct x) from t1) select 1 from w;
 ?column? 
----------
        1
(1 row)
```

When distinct agg is used in subquery and outside query have another relation
```
dev=> with w as (select min (distinct 1) from t1) select 1 from w,t2;
server closed the connection unexpectedly
        This probably means the server terminated abnormally
        before or while processing the request.
The connection to the server was lost. Attempting reset: Failed.

dev=> with w as (select min (distinct x) from t1) select 1 from w, t1;
server closed the connection unexpectedly
        This probably means the server terminated abnormally
        before or while processing the request.
The connection to the server was lost. Attempting reset: Failed.
!> with w as (select min (distinct 1) from t1) select 1 from w;
```

[frontend-4566.log](https://github.com/singularity-data/risingwave/files/9256540/frontend-4566.log)

**Comment by kwannoel** [comment]
Shrunk it further:
```
create table t1 (x int, y int);
create table t2 (x int, y int);
explain verbose select 1 from (select min(distinct 1) from t1), t2;
```

Seems like there no two-phase agg happening for left table in join (select min(distinct 1) from t1).
Additionally there seems to be some inconsistency with base and logical's distribution.
There will also need to be some exchange step inserted somewhere.
```
    left: BatchSimpleAgg {
        base: PlanBase {
            dist: SomeShard,
            ...
        },
        logical: LogicalAgg {
            base: PlanBase {
                dist: Single,
                ...
            },
```

**Comment by liurenjie1024** [comment]
cc @st1page

**Comment by kwannoel** [comment]
Fixed by #4718:
```
dev=> create table t1 (x int, y int);
CREATE_TABLE
dev=> create table t2 (x int, y int);
CREATE_TABLE
dev=> select 1 from (select min(distinct 1) from t1), t2;
 ?column? 
----------
(0 rows)
dev=> with w as (select min (distinct 1) from t1) select 1 from w,t2;
 ?column? 
----------
(0 rows)
```
